### PR TITLE
(SMT) comment contract implementation

### PR DESCRIPTION
**Pull request by eonwarped**
Adds comment contract, ability to set up post/comment reward pool.

Notes:
* Primary key functionality added to facilitate quick lookups. Compounded Indexes added to provide efficient sorting given fixed parameters (e.g. reward pool ID in this case) 
* Trending score computation and curve computations for post exponents other than 1 or 2, and curation exponents other than 0.5 or 1 use double precision float computation.
* Downvotes and upvotes use different pools, unlike hive's implementation where downvote power can bleed into the upvote power. It also uses the older voting power system (without mana) that requires powered up stake to charge before use.

### CODE REVIEW

**Review by bt-cryptomancer — CHANGES_REQUESTED** [review]
Mostly looks good, I didn't check the math carefully will assume you've tested all that thoroughly, just have some questions about some of the DB queries (see comments).  Also, you've got some lint errors to fix.

How has performance been in your test environment?  Is your node able to keep up with blocks OK, processing all the votes & comments, without lagging behind at all?

File: contracts/comments.js [hunk 1]
```diff
@@ -0,0 +1,808 @@
...
+    jsonMetadata,
+    parentAuthor,
+    parentPermlink,
+  } = payload;
+
+  // Check if it is a reply, and inherit the settings
+  // from the parent.
+  if (parentAuthor && parentPermlink) {
+    const parentAuthorperm = `@${parentAuthor}/${parentPermlink}`;
+    const parentPosts = await api.db.find('posts', { authorperm: parentAuthorperm });
```
[the remark this answers is not in the document]
Comment (eonwarped) [reply contracts/comments.js:545]: Yes, there is a cap that should be well less than 1000. I should probably add the cap explicitly as a param though.  I am planning on setting it to 20. (20 smt's linked to a post at most)

File: contracts/comments.js [hunk 2]
```diff
@@ -0,0 +1,808 @@
...
+    if (parentPosts && parentPosts.length > 0) {
+      return parentPosts.map(p => p.rewardPoolId);
+    }
+    return [];
+  }
+  // Check metadata for tags / parent permlink
+  // for community.
+  if (jsonMetadata && jsonMetadata.tags && Array.isArray(jsonMetadata.tags) && jsonMetadata.tags.every(t => typeof t === 'string')) {
+    const tagRewardPools = await api.db.find('rewardPools', { 'config.tags': { $in: jsonMetadata.tags } });
```
[the remark this answers is not in the document]
Comment (eonwarped) [reply contracts/comments.js:554]: same

File: contracts/comments.js [hunk 3]
```diff
@@ -0,0 +1,808 @@
...
+
+  if (!api.assert(api.sender === 'null', 'can only vote with voting op')) return;
+
+  if (!api.assert(Number.isInteger(weight) && weight >= -10000 && weight <= 10000,
+    'weight must be an integer from -10000 to 10000')) return;
+
+  const blockDate = new Date(`${api.hiveBlockTimestamp}.000Z`);
+  const timestamp = blockDate.getTime();
+  const authorperm = `@${author}/${permlink}`;
+  const posts = await api.db.find('posts', { authorperm });
```
[the remark this answers is not in the document]
Comment (eonwarped) [reply contracts/comments.js:801]: same

File: contracts/comments.js [hunk 4]
```diff
@@ -0,0 +1,808 @@
...
+    permlink,
+    maxAcceptedPayout,
+    beneficiaries,
+  } = payload;
+
+  // Node enforces author / permlinks from Hive. Check that sender is null.
+  if (!api.assert(api.sender === 'null', 'action must use commentOptions operation')) return;
+  const authorperm = `@${author}/${permlink}`;
+
+  const existingPosts = await api.db.find('posts', { authorperm });
```
[the remark this answers is not in the document]
Comment (eonwarped) [reply contracts/comments.js:633]: same

File: contracts/comments.js [hunk 5]
```diff
@@ -0,0 +1,808 @@
...
+      };
+      await api.db.insert('posts', post);
+      api.emit('newComment', { rewardPoolId, symbol: rewardPool.symbol });
+    }
+  }
+};
+
+/*
+ * {
+                contractName: 'comments',
```
[the remark this answers is not in the document]
Comment (eonwarped) [reply contracts/comments.js:610]: removed

**Comment by eonwarped** [comment]
> Mostly looks good, I didn't check the math carefully will assume you've tested all that thoroughly, just have some questions about some of the DB queries (see comments). Also, you've got some lint errors to fix.
> 
> How has performance been in your test environment? Is your node able to keep up with blocks OK, processing all the votes & comments, without lagging behind at all?

Performance has been good, but I need to add more pools to my test to confirm.

**Review by bt-cryptomancer — APPROVED** [review]
Changes look good, will wait to merge until witness review period is finished.

**Review by bt-cryptomancer — APPROVED** [review]
Latest changes look good.

**Review by bt-cryptomancer — DISMISSED** [review]
Latest changes look good.

**Review by bt-cryptomancer — APPROVED** [review]
Approved.

**Review by bt-cryptomancer — APPROVED** [review]
Latest changes look good.

**Review by bt-cryptomancer — APPROVED** [review]
updateFee change from 100 BEE to 20 BEE looks good.

**Pull request merged**